tests/training/test_dp_linear.py

```
@@ -0,0 +1,47 @@
+    assert set(result) == {layer.weight, layer.bias}
+    torch.testing.assert_close(result[layer.weight], torch.einsum("ni,nj->nij", backprops, activation))
+    torch.testing.assert_close(result[layer.bias], backprops)
+
+
+def test_contract_tensor_rejects_non_tensor_result(monkeypatch: pytest.MonkeyPatch):
+    def fake_contract(*_args, **_kwargs) -> int:
+        return 1
+
+    monkeypatch.setattr(linear_mod, "contract", fake_contract)
+
+    with pytest.raises(TypeError, match="expected opt_einsum.contract"):
```

Ruff (0.15.18)

[warning] Pattern passed to `match=` contains metacharacters but is neither escaped nor raw

(RUF043)